Add option to specify resampling method in load_img

ahojnnes opened the pull request:
Otherwise, nearest neighbor interpolation is used, which produces awful results. I consider this as a bug. In any case, the old behavior can be recovered using ``load_img(..., resample=PIL.Image.NEAREST)``.

myutwo150 commented on keras/preprocessing/image.py:
```diff
@@ -302,14 +302,20 @@ def img_to_array(img, data_format=None):
+            of `PIL.Image.NEAREST`, `PIL.Image.BOX`, `PIL.Image.BILINEAR`,
+            `PIL.Image.HAMMING`, `PIL.Image.BICUBIC` or `PIL.Image.LANCZOS`. If
+            omitted, or if the image has mode "1" or "P",
+            it is set `PIL.Image.NEAREST`.
```
Looks like it'll be `PIL.Image.BILINEAR` if omitted?

ahojnnes commented:
You are right, of course. Fixed now.

fchollet commented:
For the consistency with the broader API, it should be specified as a lowercase string, rather than as a constant from `PIL.Image`. Also "resampling" would probably be a better name?

ahojnnes commented:
@fchollet I addressed your comment.

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -302,14 +314,19 @@ def img_to_array(img, data_format=None):
 
-def load_img(path, grayscale=False, target_size=None):
+def load_img(path, grayscale=False, target_size=None,
+             interpolation="bilinear"):
```
Use `'` as quote char for consistency

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -302,14 +314,19 @@ def img_to_array(img, data_format=None):
         target_size: Either `None` (default to original size)
             or tuple of ints `(img_height, img_width)`.
+        interpolation: Interpolation method used to resample the image if the
+            target size does not equal the loaded image. Supported methods are
```
"is different from that of the loaded image"

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -330,7 +347,10 @@ def load_img(path, grayscale=False, target_size=None):
         if img.size != hw_tuple:
-            img = img.resize(hw_tuple)
+            if interpolation not in _PIL_INTERPOLATION_METHODS:
+                raise ValueError('Invalid interpolation method specified')
```
The ValueError should be listed in the docstring. The error message should specify what was passed, and the list of values expected instead.

ahojnnes commented:
@fchollet I addressed your comments. In addition, I added a more fine-grain check for the available interpolation methods. Only newer pillow versions support "box" and "hamming".

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -25,6 +25,22 @@
+        'bicubic': pil_image.BICUBIC,
+    }
+    # These methods were only introduced in version 3.4.0 (2016).
+    if hasattr(PIL.Image, "HAMMING"):
```
Please use consistent quote chars. If you know at which PIL version these attributes were introduce, prefer doing a version number check instead of repeated `hasattr` calls.

ahojnnes replied:
Due to the various forks of PIL and hence different versioning schemes, I'd say it is safer to use this explicit test in this case.

fchollet commented on keras/preprocessing/image.py:
```diff
@@ -328,9 +352,16 @@ def load_img(path, grayscale=False, target_size=None):
-            img = img.resize(hw_tuple)
+        width_height_tuple = (target_size[1], target_size[0])
+        if img.size != width_height_tuple:
+            if interpolation not in _PIL_INTERPOLATION_METHODS:
```
Please add a unit test for this feature, include a check for the ValueError.

ahojnnes replied:
Are there any test images shipped with keras that could be used for this test? Otherwise, it is not possible to get to this point without a failure before getting to this point.

fchollet replied:
We haves test code that creates test images and write them in a temporary folder. You can check out existing tests for `preprocessing/image`.

fchollet commented:
Re-running tests.

fchollet commented:
Re-running tests, again.

ahojnnes commented:
@fchollet Added unit tests to test the correct behavior of the interpolation and the `load_img` function itself, since I haven't found any other tests.
